dialogue-core/src/main/java/com/palantir/dialogue/core/CautiousIncreaseAggressiveDecreaseConcurrencyLimiter.java

```diff
@@ -64,10 +64,15 @@
      * */
     Optional<Permit> acquire() {
         int currentInFlight = getInflight();
-        if (currentInFlight >= getLimit()) {
-            return Optional.empty();
+
+        // We don't want to hand out a permit if there are 4 inflight and a limit of 4.1, as this will immediately send
+        // our inflight number to 5, which is clearly above the limit.  Instead, we wait until there is capacity for
+        // one whole request before handing out a permit. In the worst-case scenario with zero inflight and a limit of
+        // 1, we'll still hand out a permit
+        if (currentInFlight <= getLimit() - 1) {
+            return Optional.of(createToken());
         }
-        return Optional.of(createToken());
+        return Optional.empty();
```

Comment:
There's a race between our check, and the increment in createToken. Not sure if it's worth fixing, but could be done by deconstructing the increment into a cas loop, or optimistically incrementing and decrementing if the value failed our check (though this approach could impact metrics depending when they're collected, and should be protected by an initial pessimistic check to avoid unnecessary cas when everything is limited).

Reply:
In that case we'd still race against changes to the limit, which isn't necessarily worth thinking about.

Reply:
Yeah to be honest I'm ok with all of this being quite approximate - the concurrency limiter is always just trying to infer a sensible rate limit and even after we've handed out a permit we go through a few more layers before a network request actually gets out the door!